providers/obs/obs.go

```diff
@@ -0,0 +1,407 @@
+	name   string
+}
+
+func NewBucket(logger log.Logger, conf []byte) (*Bucket, error) {
+	config, err := parseConfig(conf)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing cos configuration")
+	}
+
+	return NewBucketWithConfig(logger, config)
+}
+
+func parseConfig(conf []byte) (Config, error) {
+	config := DefaultConfig
+	if err := yaml.UnmarshalStrict(conf, &config); err != nil {
+		return Config{}, err
+	}
+
+	return config, nil
+}
+
+func NewBucketWithConfig(logger log.Logger, config Config) (*Bucket, error) {
+	if err := config.validate(); err != nil {
+		return nil, errors.Wrap(err, "validate obs config err")
+	}
+
+	rt, err := exthttp.DefaultTransport(config.HTTPConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "get http transport err")
+	}
+
+	client, err := obs.New(config.AccessKey, config.SecretKey, config.Endpoint, obs.WithHttpTransport(rt))
+	if err != nil {
+		return nil, errors.Wrap(err, "initialize obs client err")
+	}
+
+	bkt := &Bucket{
+		logger: logger,
+		client: client,
+		name:   config.Bucket,
+	}
+	return bkt, nil
+}
+
+// Name returns the bucket name for the provider.
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+// Delete removes the object with the given name
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	input := &obs.DeleteObjectInput{Bucket: b.name, Key: name}
+	_, err := b.client.DeleteObject(input)
+	return err
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	size, err := objstore.TryToGetSize(r)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to get size apriori to upload %s", name)
+	}
+
+	if size <= 0 {
+		return errors.New("object size must be provided")
+	}
+	if size <= MinMultipartUploadSize {
+		err := b.putObjectSingle(name, r)
+		if err != nil {
+			return err
+		}
+	} else {
+		initOutput, err := b.initiateMultipartUpload(name)
+		if err != nil {
+			return err
+		}
+
+		uploadId := initOutput.UploadId
+		defer func() {
+			if err != nil {
+				if _, err := b.client.AbortMultipartUpload(&obs.AbortMultipartUploadInput{
+					UploadId: uploadId,
+					Bucket:   b.name,
+					Key:      name,
+				}); err != nil {
+					err = errors.Wrap(err, "failed to abort multipart upload")
+					return
+				}
+			}
+		}()
+		parts, err := b.multipartUpload(size, name, uploadId, r)
+		if err != nil {
+			return err
+		}
+
+		_, err = b.client.CompleteMultipartUpload(&obs.CompleteMultipartUploadInput{
+			Bucket:   b.name,
+			Key:      name,
+			UploadId: uploadId,
+			Parts:    parts,
+		})
+		if err != nil {
+			return errors.Wrap(err, "fail to complete multipart upload")
+		}
+	}
+	return nil
+}
+
+func (b *Bucket) putObjectSingle(key string, body io.Reader) error {
+	input := &obs.PutObjectInput{}
+	input.Bucket = b.name
+	input.Key = key
+	input.Body = body
+	_, err := b.client.PutObject(input)
+	return errors.Wrap(err, "fail to upload object")
```

Comment:
If find doing `if err != nil { return errors.Wrap(...) }` to be a readable a bit better rather than returning directly `errors.Wrap`.

Reply:
You are right,I will fix it